### PR TITLE
[WIP] Capture metaproj in binlog (#3402)

### PR DESCRIPTION
**Pull request by ccastanedaucf**
Currently succeeds in collecting ```.metaproj``` files into ```.binlog``` when ```MSBUILDEMITSOLUTION=1```. Example below shows opening the ```.binlog``` from MSBuildStructuredLogViewer where```ConsoleApp1``` has a build dependency on ```ClassLib1```.

metaproj

### CODE REVIEW

File: src/Framework/MetaProjectGeneratedEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Build.Framework
+{
+    public class MetaProjectGeneratedEventArgs : BuildMessageEventArgs
```
Comment (ccastanedaucf) [inline src/Framework/MetaProjectGeneratedEventArgs.cs]: Left this as a skeleton since I'm looking for feedback if there's an existing EventArgs class I should be using for this instead of creating a new one.

**Comment by ccastanedaucf** [comment]
@dotnet-bot test Ubuntu16.04 Build for CoreCLR please

**Comment by ccastanedaucf** [comment]
@dotnet-bot test OSX10.13 Build for CoreCLR please

File: src/Framework/MetaProjectGeneratedEventArgs.cs (hunk 1 above)
Comment (rainersigwald) [reply src/Framework/MetaProjectGeneratedEventArgs.cs]: I think creating a new one is reasonable.

File: src/Build/Construction/Solution/SolutionProjectGenerator.cs [hunk 2]
```diff
@@ -912,6 +915,16 @@ private ProjectInstance CreateTraversalInstance(string wrapperProjectToolsVersio
             return traversalInstance;
         }
 
+        private void LogMetaProjectGeneratedEvent(string fullPath)
```
[the remark this answers is not in the document]
Comment (ccastanedaucf) [reply src/Build/Construction/Solution/SolutionProjectGenerator.cs]: It's passed the path to each generated ```.metaproject``` under ```CreateSolutionProject()```, and the single ```.metaproject.tmp``` under ```CreateTraversalInstance()```

**Comment by ccastanedaucf** [comment]
I placed it in the screencap but I forgot to mention ```.metaproj.tmp``` is also captured here.

The only caveat I see with capturing them all the time is that ```ProjectImportsCollector``` in ```BinaryLogger``` is going to need the ```.metaproj``` files written to disk beforehand, otherwise if ```MSBUILDEMITSOLUTION=0``` it's never going to capture anything. Only way I see around always creating the file is passing each ```ProjectRootElement``` into ```MetaProjectGeneratedEventArgs``` then  letting the ```BinaryLogger``` create the file if it hasn't been generated, but that kind of seems out of it's responsibility.

**Comment by rainersigwald** [comment]
Yeah, that's what I was thinking. I wonder if it would cause problems for a multiproc scenario though, since the project root element (and instance) would have to be serialized back. I guess we can tackle that later.

**Comment by cdmihai** [comment]
Is the new `MetaProjectGeneratedEventArgs` event really needed, as opposed to using the existing `ProjectEvaluationStartedEventArgs`? When the project file collection was implemented, I don't think the evaluation started / ended events were there to be used. Maybe we can even get rid of relying on `ProjectStartedEventArgs`, and just use `ProjectEvaluationStartedEventArgs` and `ProjectImportedEventArgs` to capture all the files that get looked at.

**Comment by ccastanedaucf** [comment]
Got it emitting the event in all instances. Found passing the metaproj objects didn't work so well since ```ProjectInstance``` seems to get mutated in multiple places afterwards which affects the serialization output, and for ```metaproj.tmp``` some data is lost when evaluating to a project instance so we need the state at that moment. Tried various methods of passing the metaproj data but perf tests pretty big regressions on memory usage.

But just writing to disk temporarily with the flag off doesn't seem to make a huge impact on performance, plus the binary logger is able to stream the data as it does with project imports. So here's the perf tests for that below.

Also tests are going to fail again from having a null logger reference.

### DesignTimeBuild Time (ms) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 36.6261 -> 36.345 (-0.767%)
DotnetWebProject | :red_circle: | yes | 226.1276 -> 232.8492 (2.972%)
DotnetMvcProject | :ok_hand: | no | 227.8292 -> 228.9814 (0.506%)
Picasso | :white_check_mark: | yes | 1594.3534 -> 1590.847 (-0.22%)
SmallP2POldCsproj | :ok_hand: | no | 69.5209 -> 69.57 (0.071%)
SmallP2PNewCsproj | :red_circle: | yes | 634.3607 -> 637.6251 (0.515%)
LargeP2POldCsproj | :white_check_mark: | yes | 11733.5193 -> 11655.6611 (-0.664%)
OrchardCore | :ok_hand: | no | 49903.1406 -> 49949.1597 (0.092%)

### DesignTimeBuild Memory (bytes) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 6274630 -> 6279506 (0.078%)
DotnetWebProject | ::ok_hand: | no | 24123050 -> 23390010 (-3.039%)
DotnetMvcProject | :ok_hand: | no | 22116857 -> 22990710 (3.951%)
Picasso | :white_check_mark: | yes | 186972661 -> 178562502 (-4.498%)
SmallP2POldCsproj | ::ok_hand: | no | 8614571 -> 8608660 (-0.069%)
SmallP2PNewCsproj | :ok_hand: | no | 89834691 -> 90417988 (0.649%)
LargeP2POldCsproj | ::ok_hand: | no | 1166614054 -> 1158014712 (-0.737%)
OrchardCore | :white_check_mark: | yes | 2717116171 -> 2612868480 (-3.837%)

### SerialEvaluationIsolatedContext Time (ms) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 35.8413 -> 35.3005 (-1.509%)
DotnetWebProject | :white_check_mark: | yes | 49.8859 -> 49.7245 (-0.324%)
DotnetMvcProject | ::ok_hand: | no | 56.1874 -> 56.1251 (-0.111%)
Picasso | :ok_hand: | no | 312.8652 -> 313.3025 (0.14%)
SmallP2POldCsproj | ::ok_hand: | no | 53.0437 -> 53.0287 (-0.028%)
SmallP2PNewCsproj | :red_circle: | yes | 205.0586 -> 207.7923 (1.333%)
LargeP2POldCsproj | :white_check_mark: | yes | 929.0208 -> 923.0702 (-0.641%)
OrchardCore | ::ok_hand: | no | 3361.7828 -> 3356.1962 (-0.166%)

### SerialEvaluationIsolatedContext Memory (bytes) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 5726103 -> 5719864 (-0.109%)
DotnetWebProject | :white_check_mark: | yes | 7743319 -> 7734272 (-0.117%)
DotnetMvcProject | :white_check_mark: | yes | 8468958 -> 8453648 (-0.181%)
Picasso | ::ok_hand: | no | 11319168 -> 10999293 (-2.826%)
SmallP2POldCsproj | ::ok_hand: | no | 7064488 -> 7062922 (-0.022%)
SmallP2PNewCsproj | ::ok_hand: | no | 16037744 -> 15037786 (-6.235%)
LargeP2POldCsproj | ::ok_hand: | no | 40660372 -> 39730102 (-2.288%)
OrchardCore | ::ok_hand: | no | 66031067 -> 61743716 (-6.493%)

### SerialEvaluationSharedContextSecondRun Time (ms) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 29.9967 -> 29.9782 (-0.062%)
DotnetWebProject | :red_circle: | yes | 37.2908 -> 37.5104 (0.589%)
DotnetMvcProject | :ok_hand: | no | 39.9996 -> 40.0399 (0.101%)
Picasso | :white_check_mark: | yes | 227.413 -> 226.6705 (-0.326%)
SmallP2POldCsproj | :ok_hand: | no | 42.9467 -> 42.9864 (0.092%)
SmallP2PNewCsproj | :white_check_mark: | yes | 134.3253 -> 133.0591 (-0.943%)
LargeP2POldCsproj | :red_circle: | yes | 688.5057 -> 696.3853 (1.144%)
Generated_100_100_v150 | :red_circle: | yes | 1181.8339 -> 1185.1449 (0.28%)
OrchardCore | :red_circle: | yes | 1641.6273 -> 1650.5991 (0.547%)
Roslyn | :ok_hand: | no | 2864.7094 -> 2869.3471 (0.162%)
WebLargeCore | :white_check_mark: | yes | 1712.4508 -> 1695.2265 (-1.006%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 5567643 -> 5575508 (0.141%)
DotnetWebProject | :red_circle: | yes | 6919179 -> 6920319 (0.016%)
DotnetMvcProject | :red_circle: | yes | 7378854 -> 7380015 (0.016%)
Picasso | :red_circle: | yes | 36668313 -> 36689583 (0.058%)
SmallP2POldCsproj | :red_circle: | yes | 6984795 -> 6992478 (0.11%)
SmallP2PNewCsproj | :red_circle: | yes | 23026288 -> 23043522 (0.075%)
LargeP2POldCsproj | ::ok_hand: | no | 37819065 -> 37272616 (-1.445%)
Generated_100_100_v150 | :red_circle: | yes | 46438705 -> 46696482 (0.555%)
OrchardCore | ::ok_hand: | no | 62251107 -> 61003247 (-2.005%)
Roslyn | :ok_hand: | no | 92421524 -> 94205796 (1.931%)
WebLargeCore | :red_circle: | yes | 51837521 -> 56249550 (8.511%)

### SerialEvaluationSharedContext Time (ms) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 34.6246 -> 34.5696 (-0.159%)
DotnetWebProject | :red_circle: | yes | 48.694 -> 50.0102 (2.703%)
DotnetMvcProject | :ok_hand: | no | 55.7387 -> 55.7447 (0.011%)
Picasso | ::ok_hand: | no | 258.3352 -> 257.805 (-0.205%)
SmallP2POldCsproj | :white_check_mark: | yes | 47.5375 -> 47.0647 (-0.995%)
SmallP2PNewCsproj | :ok_hand: | no | 163.6252 -> 164.826 (0.734%)
LargeP2POldCsproj | :red_circle: | yes | 739.4428 -> 748.0903 (1.169%)
Generated_100_100_v150 | :red_circle: | yes | 1220.556 -> 1224.6891 (0.339%)
OrchardCore | :ok_hand: | no | 3035.9925 -> 3040.9852 (0.164%)
Roslyn | :red_circle: | yes | 3434.4311 -> 3442.4076 (0.232%)
WebLargeCore | :white_check_mark: | yes | 2246.8132 -> 2223.6065 (-1.033%)

### SerialEvaluationSharedContext Memory (bytes) (src\msbuild.master vs src\msbuild.chcasta)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 5788910 -> 5788867 (-0.001%)
DotnetWebProject | :white_check_mark: | yes | 7793321 -> 7784775 (-0.11%)
DotnetMvcProject | :white_check_mark: | yes | 8510965 -> 8501368 (-0.113%)
Picasso | :ok_hand: | no | 12216643 -> 12356449 (1.144%)
SmallP2POldCsproj | :red_circle: | yes | 7055765 -> 7066977 (0.159%)
SmallP2PNewCsproj | :white_check_mark: | yes | 19299584 -> 16530177 (-14.35%)
LargeP2POldCsproj | :ok_hand: | no | 37492382 -> 38288598 (2.124%)
Generated_100_100_v150 | :red_circle: | yes | 46934231 -> 47818337 (1.884%)
OrchardCore | :ok_hand: | no | 67120656 -> 70771017 (5.439%)
Roslyn | :ok_hand: | no | 90160994 -> 92736015 (2.856%)
WebLargeCore | :ok_hand: | no | 59067591 -> 61608146 (4.301%)

**Comment by ccastanedaucf** [comment]
Also @cdmihai I took a look at the start / ended events and it looks like none of them are instantiated outside of tests and ```LoggingServiceLogMethods``` and mostly used in loggers so I don't think they are there to be used either. The other event args pop up in multiple places though.

File: src/Build/Construction/Solution/SolutionProjectGenerator.cs [hunk 3]
```diff
@@ -912,6 +907,18 @@ private ProjectInstance CreateTraversalInstance(string wrapperProjectToolsVersio
             return traversalInstance;
         }
 
+        private void EmitMetaProject(ProjectRootElement metaProject, string metaProjectPath)
+        {
+            string filePath = Environment.GetEnvironmentVariable("MSBuildEmitSolution") == null ? Path.GetTempFileName() : metaProjectPath;
+            metaProject.Save(filePath);
```
Comment (rainersigwald) [inline src/Build/Construction/Solution/SolutionProjectGenerator.cs]: Would it be reasonable to use the `ProjectRootElement.Save(TextWriter)` overload here to avoid the physical file write?

**Comment by cdmihai** [comment]
> Also @cdmihai I took a look at the start / ended events and it looks like none of them are instantiated outside of tests and LoggingServiceLogMethods and mostly used in loggers so I don't think they are there to be used either. The other event args pop up in multiple places though.

The evaluation started / ended events already get triggered for each evaluation, so you wouldn't have to trigger them, just listen for them in the binary logger file scraper. The benefit here is that you wouldn't have to introduce a new event.

File: src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs [hunk 4]
```diff
@@ -138,8 +147,9 @@ public void EmitToolsVersionAttributeToInMemoryProject9()
                 ";
 
             SolutionFile solution = SolutionFile_Tests.ParseSolutionHelper(solutionFileContents);
+            var buildEventContext = new BuildEventContext(0, 0, BuildEventContext.InvalidProjectContextId, 0);
```
Comment (ccastanedaucf) [inline src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs]: Wondering if anyone else has more insight into ```BuildEventContext```. All the changed tests were failing unless the third argument to the constructor is ```BuildEventContext.InvalidProjectContextId```, and failing by Windows bringing up a crash window without a useful stack trace. Noticed this because all the passing tests happened to be passing ```BuildEventContext.Invalid``` to ```Generate()```.

File: src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs [hunk 5]
```diff
@@ -1201,14 +1221,14 @@ public void ToolsVersionOverrideCausesToolsetRedirect()
 
             ProjectInstance[] instances = null;
             SolutionFile solution = SolutionFile_Tests.ParseSolutionHelper(solutionFileContents);
-            BuildEventContext buildEventContext = new BuildEventContext(0, 0, 0, 0);
+            var buildEventContext = new BuildEventContext(0, 0, BuildEventContext.InvalidProjectContextId, 0);
```
Comment (cdmihai) [inline src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs]: nit: you can extract this out in a field that gets instantiated in the constructor. This way you only have to define it once.

File: src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs (hunk 4 above)
Comment (cdmihai) [reply src/Build.UnitTests/Construction/SolutionProjectGenerator_Tests.cs]: Could be this check which is executed only when ProjectContextId is valid: https://github.com/Microsoft/msbuild/blob/2a012e653766eb261b09b29b3106f4eb57a7f61d/src/Build/BackEnd/Components/Logging/LoggingService.cs#L1457

You should look at the stack trace and see where the exception is coming from, as it could point to further issues. As far as I can remember, when in Debug, MSBuild sometimes emits that crash window to let you easily attach a debugger. If you click Retry (or abort, can't remember), it will spawn the debugger attach window.

File: src/Build/Construction/Solution/SolutionProjectGenerator.cs [hunk 6]
```diff
@@ -911,7 +911,8 @@ private void EmitMetaProject(ProjectRootElement metaProject, string metaProjectP
         {
             string filePath = Environment.GetEnvironmentVariable("MSBuildEmitSolution") == null ? Path.GetTempFileName() : metaProjectPath;
             metaProject.Save(filePath);
-            var eventArgs = new MetaProjectGeneratedEventArgs(metaProjectPath)
+            string message = ResourceUtilities.GetResourceString("MetaprojectGenerated");
+            var eventArgs = new MetaProjectGeneratedEventArgs(metaProjectPath, message)
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply src/Build/Construction/Solution/SolutionProjectGenerator.cs]: Looks alright to me. The events get fired before the sln events, which is representative of reality.

@KirillOsenkov, the creator of the log viewer, for opinions :)

**Comment by ccastanedaucf** [comment]
New commits start at 
432b8ae
Had a weird rebase issue with localization changes from master so it threw off the history a bit.

File: src/Build/Construction/Solution/SolutionProjectGenerator.cs [hunk 7]
```diff
@@ -912,6 +907,29 @@ private ProjectInstance CreateTraversalInstance(string wrapperProjectToolsVersio
             return traversalInstance;
         }
 
+        private void EmitMetaproject(ProjectRootElement metaproject, string path)
+        {
+            if (Environment.GetEnvironmentVariable("MSBuildEmitSolution") != null)
```
Comment (AndyGerlicher) [inline src/Build/Construction/Solution/SolutionProjectGenerator.cs]: This will get called once per project now instead of once. Ideally we could move this to `Traits` and it would be cached so it doesn't matter.

**Comment by AndyGerlicher** [comment]
@panopticoncentral just FYI we're adding to your interface for 15.8. We didn't add a new interface because it hasn't been released non-preview yet.

**Comment by AndyGerlicher** [comment]
Closing in favor of #3498
